### PR TITLE
365 rework tc status magic

### LINKED ISSUE #366: crypto_tc.c  Rework status code magic.
Note that we should go back to lines 1974 - 81  And resolve the error codes - Will probably require a little bit of code rework inside of Crypto_Process_Extended_Procedure_Pdu().

_Originally posted by @rjbrown2 in https://github.com/nasa/CryptoLib/issues/365#issuecomment-2533166166_

### PR DESCRIPTION
**Pull request by dccutrig**
Resolves #366 (got the branch number off by one)

### CODE REVIEW

File: src/core/crypto.c [hunk 1]
```diff
@@ -772,99 +773,126 @@ int32_t Crypto_Get_Managed_Parameters_For_Gvcid(uint8_t tfvn, uint16_t scid, uin
...
         status = CRYPTO_LIB_ERR_NULL_BUFFER;
     }
+
+    // Validate correct SA for EPs
+    bool valid_ep_sa = (tc_sdls_processed_frame->tc_sec_header.spi == SPI_MIN) || (tc_sdls_processed_frame->tc_sec_header.spi == SPI_MAX);
     if (status == CRYPTO_LIB_SUCCESS)
     {
         if (crypto_config.has_pus_hdr == TC_HAS_PUS_HDR)
         {
+            // Check for speciic App ID for EPs - the CryptoLib Apid in this case
```
Comment (Donnie-Ice) [inline src/core/crypto.c:790]: Typo

**Review by Donnie-Ice — APPROVED** [review]

File: src/core/crypto.c [hunk 2]
```diff
@@ -21,6 +21,7 @@
 */
 #include "crypto.h"
 #include <string.h>
+#include <stdbool.h>
```
Comment (jlucas9) [inline src/core/crypto.c:24]: Do we use bools anywhere else or should we just use a uint8_t and check true / false against pound def?

File: src/core/crypto.c [hunk 3]
```diff
@@ -772,99 +773,126 @@ int32_t Crypto_Get_Managed_Parameters_For_Gvcid(uint8_t tfvn, uint16_t scid, uin
...
+                    sdls_frame.hdr.shdr = (tc_sdls_processed_frame->tc_pdu[0] & 0x08) >> 3;
+                    sdls_frame.hdr.appID =
+                        ((tc_sdls_processed_frame->tc_pdu[0] & 0x07) << 8) | tc_sdls_processed_frame->tc_pdu[1];
+                    sdls_frame.hdr.seq = (tc_sdls_processed_frame->tc_pdu[2] & 0xC0) >> 6;
+                    sdls_frame.hdr.pktid =
+                        ((tc_sdls_processed_frame->tc_pdu[2] & 0x3F) << 8) | tc_sdls_processed_frame->tc_pdu[3];
+                    sdls_frame.hdr.pkt_length =
+                        (tc_sdls_processed_frame->tc_pdu[4] << 8) | tc_sdls_processed_frame->tc_pdu[5];
+
+                    // CCSDS PUS
```
[the remark this answers is not in the document]
Comment (dccutrig) [reply src/core/crypto.c]: I believe this is resolved.

File: src/core/crypto.c (hunk 2 above)
Comment (dccutrig) [reply src/core/crypto.c]: Resolved

**Comment by rjbrown2** [comment]
Do we happen to have a test w/o the PUS header?  I know we have the one that caught this initial issue, but did we add another for one that doesn't have one? (Just thinking aloud)

**Comment by dccutrig** [comment]
@rjbrown2 I have tests now with and without PUS header, and some bounds checking. See what ya think

File: src/core/crypto.c [hunk 4]
```diff
@@ -802,67 +811,126 @@ int32_t Crypto_Process_Extended_Procedure_Pdu(TC_t *tc_sdls_processed_frame, uin
...
-                status = CRYPTO_LIB_ERR_SDLS_EP_NOT_BUILT;
-                #endif //CRYPTO_EPROC
+            
+            // Determine type of PDU
+            status = Crypto_PDU(ingest, tc_sdls_processed_frame);
             }
+            // Received EP PDU on invalid SA
+            else
+            {
+                printf(KRED "Received EP PDU on invalid SA! SPI %d\n" RESET, tc_sdls_processed_frame->tc_sec_header.spi);
```
Comment (Donnie-Ice) [inline src/core/crypto.c:888]: add `#ifdef DEBUG` to all prints

File: src/core/crypto.c [hunk 5]
```diff
@@ -802,67 +811,126 @@ int32_t Crypto_Process_Extended_Procedure_Pdu(TC_t *tc_sdls_processed_frame, uin
...
-        {
-            // TODO - Process SDLS PDU with Packet Layer without PUS_HDR
+                // Determine type of PDU
+                status = Crypto_PDU(ingest, tc_sdls_processed_frame);
+            }
+            #else // Received an EP command without EPs being built
+#ifdef CCSDS_DEBUG
+            printf(KRED "PDU DEBUG %s %d\n" RESET, __FILE__, __LINE__);
+#endif
+            valid_ep_sa = valid_ep_sa; // Suppress build error
```
Comment (Donnie-Ice) [inline src/core/crypto.c:931]: line 931 is duplicated on line 893

Comment (Donnie-Ice) [reply src/core/crypto.c:931]: actually, might be due to if statement

File: src/core/crypto_tc.c [hunk 6]
```diff
@@ -1967,17 +1967,7 @@ int32_t Crypto_TC_ProcessSecurity_Cam(uint8_t *ingest, int *len_ingest, TC_t *tc
...
-            status = Crypto_Process_Extended_Procedure_Pdu(tc_sdls_processed_frame, ingest);
-        }
-        else  
-        {
-            // Some Magic here to log that an inappropriate SA was attempted to be used for EP
-            status = CRYPTO_LIB_ERR_SPI_INDEX_OOB;  
-            mc_if->mc_log(status);
-            status = CRYPTO_LIB_SUCCESS;
-        }
+        status = Crypto_Process_Extended_Procedure_Pdu(tc_sdls_processed_frame, ingest);
```
Comment (jlucas9) [inline src/core/crypto_tc.c:1970]: Do we still need the SPI_MIN and SPI_MAX check above?

**Review by rjbrown2 — APPROVED** [review]

File: src/core/crypto.c (hunk 5 above)
Comment (dccutrig) [reply src/core/crypto.c:931]: Due to if

File: src/core/crypto_tc.c (hunk 6 above)
Comment (dccutrig) [reply src/core/crypto_tc.c:1970]: Yes

**Pull request merged**